rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,393 @@
+and perform unneeded inter-region transfers.  By transmitting a list of
+requests which allow the uploading to happen, we can instruct the uploader to
+upload directly to the existing S3 infrastructure, but perform all signing
+within this service.
+
+A generalised request will look like:
+
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in
+  the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key ->
+  list of strings)
+* All query string, path, fragments are contained in a fully formed `url`
+  property
+
+### Origins
+Where present, the `origin` rest parameter in this service will be either an
+IPv4 address, IPv6 address, or an identifier.  This parameter will specify the
+source location of the ultimately originating request to upload.  If the origin
+is an IPv4 or IPv6, a mapping of the address and IP will occur to find the
+backing storage.  If the origin is not an IPv4 or IPv6 address and is identical
+to a set of known and mappable identifiers (e.g.  `s3_us-west-2`), then that
+will be used to map the token to find the backing storage provider.  If the
+origin is not an IP address and not a configured identifier, an error will be
+sent back to the client.
+
+The mapping of origins to storage providers should be statically configured.
+
+The origin `all` is reserved and cannot be used.  This is to enable management
+endpoints to operate on all origins in the future.
+
+### Creation
+
+```
+PUT /objects/:name
+PATCH /objects/:name
+```
+
+The `PUT` endpoint will take an origin and name.  It will return a `200` with a
+document that contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the object in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+  parts: [{
+    sha256: 'abcd1234',
+    size: 1024,
+  }, {
+    sha256: 'abcd1234',
+    size: 1024,
+  }],
+}
+```
+
+This endpoint will return a list of HTTP requests which must all be run.
+
+The `PATCH` endpoint will be sent without a request body and will be used to
+perform any post-upload actions.  These include things like the commit step on
+an S3 multipart upload.  This endpoint would be run by the frontend service and
+not by the uploader, and the frontend service would also need a complimentary
+method for workers to call when they complete the upload.  This is what is
+currently done with the `blob` storage type in the queue.
```

Comment:
The use of the HTTP `PATCH` method implies adherence to RFC 5789. Adhering to the RFC provides little to no benefits and has the drawback that future behavior may be governed by spec-conforming clients. I think you should ditch `PATCH` and use `POST` instead.

Reply:
Sure, but for what it's worth I think we've used `PATCH` in other parts of other taskcluster apis.

Reply:
I don't think we use it without a body, though -- that could conceivably cause issues with HTTP implementations.